persistent/picklecache.py

```diff
@@ -229,36 +296,105 @@ def debug_info(self):
 
     # Helpers
-    def _sweep(self, target):
+
+    # Set to true when a deactivation happens in our code. For
+    # compatibility with the C implementation, we can only remove the
+    # node and decrement our non-ghost count if our implementation
+    # actually runs (broken subclasses can forget to call super; ZODB
+    # has tests for this). This gets set to false everytime we examine
+    # a node and checked afterwards. The C implementation has a very
+    # incestuous relatiounship between cPickleCache and cPersistence:
+    # the pickle cache calls _p_deactivate, which is responsible for
+    # both decrementing the non-ghost count and removing its node from
+    # the cache ring. We're trying to keep that to a minimum, but
+    # there's no way around it if we want full compatibility
+    _persistent_deactivate_ran = False
+
+    @_sweeping_ring
+    def _sweep(self, target, target_size_bytes=0):
         # lock
         node = self.ring.next
-        while node is not self.ring and self.non_ghost_count > target:
-            if node.object._p_state not in (STICKY, CHANGED):
-                node.prev.next, node.next.prev = node.next, node.prev
-                node.object = None
-                self.non_ghost_count -= 1
+        ejected = 0
+
+        while (node is not self.ring
+               and ( self.non_ghost_count > target
+                    or (target_size_bytes and self.total_estimated_size > target_size_bytes))):
+            if node.object._p_state == UPTODATE:
+                # The C implementation will only evict things that are specifically
+                # in the up-to-date state
+                self._persistent_deactivate_ran = False
+
+                # sweeping an object out of the cache should also
+                # ghost it---that's what C does. This winds up
+                # calling `update_object_size_estimation`.
+                # Also in C, if this was the last reference to the object,
+                # it removes itself from the `data` dictionary.
+                # If we're under PyPy or Jython, we need to run a GC collection
+                # to make this happen...this is only noticeable though, when
+                # we eject objects. Also, note that we can only take any of these
+                # actions if our _p_deactivate ran, in case of buggy subclasses.
+                # see _persistent_deactivate_ran
+
+                node.object._p_deactivate()
+                if (self._persistent_deactivate_ran
+                    # Test-cases sneak in non-Persistent objects, sigh, so naturally
+                    # they don't cooperate (without this check a bunch of test_picklecache
+                    # breaks)
+                    or not isinstance(node.object, _SWEEPABLE_TYPES)):
+                    ejected += 1
+                    self.__remove_from_ring(node)
             node = node.next
+        if ejected:
+            # TODO: Only do this on PyPy/Jython?
```

Comment:
In ZODB, everytime a connection is closed, it calls `incrgc` on its cache. I think it does this to reduce it down to its target size in case many extra objects were added during the course of the transaction---this in turn frees them for the Python collector to reclaim, which in C happens immediately.

The Python PickleCache implementation keeps track of the objects it is caching in a `WeakValueDictionary`. The number of objects in the cache (in this dictionary) is exposed as the `len` of the cache. Under non-refcounted implementations like PyPy, the weak references in this dictionary are only cleared when the garbage collector runs. Thus, after an `incrgc`, the `len` of the cache is incorrect for some period of time *unless* we ask the GC to run.

There is at least one test in ZODB (testConnectionSavepoint) that relies on the length of the cache being accurate after `incrgc`. It could probably be changed, but I was trying to keep the changes in ZODB minimal.